analysis/coverage_data_utils.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility functions for coverage data calculation."""
+
+import posixpath
+from collections import defaultdict
+import tempfile
+import json
+import os
+import pandas as pd
+
+from common import filestore_utils
+
+
+def get_fuzzer_benchmark_key(fuzzer: str, benchmark: str):
+    """Returns the key in coverage dict for a pair of fuzzer-benchmark."""
+    return fuzzer + ' ' + benchmark
+
+
+def get_fuzzer_filestore_path(benchmark_df, fuzzer):
+    """Gets the filestore_path for |fuzzer| in |benchmark_df|."""
+    fuzzer_df = benchmark_df[benchmark_df.fuzzer == fuzzer]
+    filestore_path = fuzzer_df.experiment_filestore.unique()[0]
+    exp_name = fuzzer_df.experiment.unique()[0]
+    return posixpath.join(filestore_path, exp_name)
+
+
+def get_covered_regions_dict(experiment_df):
+    """Combines json files for different fuzzer-benchmark pair
+    in |experiment_df| and returns a dictionary of the covered regions."""
+    covered_regions_dict = {}
+    benchmarks = experiment_df.benchmark.unique()
+    for benchmark in benchmarks:
+        benchmark_df = experiment_df[experiment_df.benchmark == benchmark]
+        fuzzers = benchmark_df.fuzzer.unique()
+        for fuzzer in fuzzers:
+            fuzzer_covered_regions = get_fuzzer_covered_regions(
+                benchmark_df, benchmark, fuzzer)
+            key = get_fuzzer_benchmark_key(fuzzer, benchmark)
+            covered_regions_dict[key] = fuzzer_covered_regions
+    return covered_regions_dict
+
+
+def get_fuzzer_covered_regions(benchmark_df, benchmark, fuzzer):
+    """Gets the covered regions for |fuzzer| in |benchmark_df| from the json
+    file in the bucket."""
+    with tempfile.TemporaryDirectory() as tmpdirname:
```

Comment:
Actually, I don't think you need tempdirectory at all here.
Can you just used NamedTemporaryFile instead?

Reply:
True, I'll change it.

Reply:
I just tried NamedTemporaryFile but I think it doesn't work since the file is open and if you copy the json file from gcs to that `NamedTemporaryFile.name` you will create a permanent file with the same path.

Reply:
OK, directory is fine.